### ISSUE TITLE
[XLA] Unintuitive behavior of jit_scope

### ISSUE THREAD

**Issue by zhuzilin** [tensorflow/tensorflow#41626]
As far as I can tell, the implementation of `MarkForCompilation` relates to `jit_scope` (`_XlaCompile` attribute) does the following steps:
1. Remove the nodes whose `_XlaCompile` attribute is set to false from the compilation candidates (using `CompilationDisallowedByXlaCompileAttr`).
2. Create a cluster for each candidate using `_XlaScope` as the scope (`absl::nullopt` otherwise).
3. Cluster the one-node clusters and set `is_xla_compile_attr_true` to true if any node in the cluster has `_XlaCompile` true.
4. Add the `kXlaClusterAttr` to node if the `is_xla_compile_attr_true` of its cluster is true (as in `ShouldCompileClusterImpl`).

The problem of the above steps is that if we want to only cluster part of the model, for example:
```python
def stage1(input):
  # stage 1 of the model

def stage2(x):
  # stage2 of the model

def loss(x):
  # ...

x = stage1(input)
with tf.xla.experimental.jit_scope():
  x = stage2(x)
l = loss(x)
train_op = optimizer.minimize(l)

# ...
```
The unlabeled part of the model may still be clustered and compiled. In practice, such situation happens a lot in the backprop part of the model, where the unlabeled loss gradient ops are clustered with the gradient ops inside `jit_scope`. I wonder if this is the expected behavior and I think that it is different from the intuitive idea that only the ops in `jit_scope` are compiled.

One walkaround to this problem is adding a global `jit_scope(compile_ops=False)`:

```python
with tf.xla.experimental.jit_scope(compile_ops=False):
  x = stage1(input)
  with tf.xla.experimental.jit_scope():
    x = stage2(x)
  l = loss(x)
  train_op = optimizer.minimize(l)
```

Following are the comparison between the origin result and the walkaround

without the global `jit_scope(compile_ops=False)`:
![image](https://user-images.githubusercontent.com/10428324/88171134-678e0780-cc51-11ea-8b96-32ad29245a35.png)

with the global `jit_scope(compile_ops=False)`:
![image](https://user-images.githubusercontent.com/10428324/88171067-4b8a6600-cc51-11ea-9ad4-7895fc830b22.png)

The red part of the timeline is the `XlaRun` op.

I think that to solve this problem, we only need to remove the unlabeled ops from the compilation candidates when global jit level is not set. It will be great if the XLA team would like to share some idea on the current behavior of `jit_scope`.

Also, is `global_jit_level` designed to be override the `jit_scope`? It seems that if `global_jit_level` is set, it will ignore the `_XlaScope` attribute and use `_XlaInternalScope` instead. In this way, if one want to compile the whole model, is he or she allowed to separate the cluster arbitrarily?

Thank you for your time on this issue :).

Gently ping @sanjoy @cheshire @tpopp .

**Comment by ravikyram** [comment]
@zhuzilin 

Request you to fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.
Please, share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

**Comment by zhuzilin** [comment]
@ravikyram 
I believe this issue should be classified into the "other" category because it mainly relates to the internal structure of XLA. And because it relates to timeline, I'm not sure it can be easily reproduced through colab and I think the content has already localized the issue... As for the tensorflow version, I think the implementation mentioned exists from at least 1.15.

**Comment by cheshire** [comment]
This is all running without autoclustering, right?

> In practice, such situation happens a lot in the backprop part of the model, where the unlabeled loss gradient ops are clustered with the gradient ops inside jit_scope

To me it actually kind of makes sense: the computation is marked as "best-effort compile", so the gradient over the computation is also "best-effort compile"?.. Am I missing something?

**Comment by cheshire** [comment]
> It seems that if global_jit_level is set, it will ignore the _XlaScope attribute and use _XlaInternalScope instead

Yes, using explicit JIT scopes when autoclustering is on does not really make sense.

Also just to set expectations now we are mostly focusing on explicit compilation with `tf.function(experimental_compile=True)` as a preferred API for using XLA in TF.

**Comment by zhuzilin** [comment]
@cheshire 
Thank you for your explanation :).  
> To me it actually kind of makes sense: the computation is marked as "best-effort compile", so the gradient over the computation is also "best-effort compile"?.. Am I missing something?

Here is where the misunderstanding emerges. The doc for `jit_scope` says:
> Enable or disable JIT compilation of operators within the scope.
> ...
> The compilation is a hint and only supported on a best-effort basis.

It may be wrongly interpreted as "XLA will try to compile the whole scope, but because XLA may not support compiling all of the scope, it will be regard as a hint." instead of "XLA will may merge the ops in the scope with ones out of it for better performance." Also, it will be a little confusing that the compilation result will be different for ops in one place if we add a `jit_scope` at the other. The current behavior will be inconvenient for those who are using `jit_scope` to debug if any part of model is unsuitable to  XLA and has led to negative optimization, because the problem part may always been clustered and compiled. I think that it will be more consistent with the design of tensorflow api that a "scope" would only have influence to things inside it.

I understand that we hope to have better performance by creating larger clusters. If there is little motivation to change the design of `jit_scope`, can I help to clarify this possible misunderstanding in the doc and provide an example for the users who only hope to compile the ops inside a certain scope?

**Comment by cheshire** [comment]
@zhuzilin For the doc, would you want to send a PR with examples? :P 

From a different angle, is `tf.function(experimental_compile=True)` unsuitable for your purposes?

**Comment by zhuzilin** [comment]
@cheshire 
I'd love to send a pr :). As for `tf.function`, many legacy models in our team are till using tf1.x... So we cannot use it for now...

**Comment by zhuzilin** [comment]
@cheshire 
There is another question relates to `jit_scope`. In `_MaybeCompile()`, we have:
```python
def _MaybeCompile(scope, op, func, grad_fn):
  """Compile the calculation in grad_fn if op was marked as compiled."""
  scope = scope.rstrip("/").replace("/", "_")
  if func is not None:
    xla_compile = func.definition.attr["_XlaCompile"].b
    xla_separate_compiled_gradients = func.definition.attr[
        "_XlaSeparateCompiledGradients"].b
    xla_scope = func.definition.attr["_XlaScope"].s.decode()
  else:
    try:
      xla_compile = op.get_attr("_XlaCompile")
      xla_separate_compiled_gradients = op.get_attr(
          "_XlaSeparateCompiledGradients")
      xla_scope = op.get_attr("_XlaScope").decode()
    except ValueError:
      return grad_fn()  # Exit early

  if not xla_compile:
    return grad_fn()  # Exit early
  ...
```
I wonder if the latter "Exit early" (`jit_scope(compile_ops=False)` won't have influence to the grad fn) is the expected feature or a bug? Thank you.
